Implementing a `/get_inventorywithsites` endpoint

raysapida opened the pull request:
- Uses the query ItemSitesQueryRq
- Add a new class `Inventorywithsites` that only deals with polling
- Using vendor_query_rs as a template for the new item_site_query_rs

raysapida committed:
Implementing an `get_inventorywithsites` endpoint

- Uses the query ItemSitesQueryRq
- Add a new class `Inventorywithsites` that only deals with polling
- Using vendor_query_rs as a template for the new item_site_query_rs

gorandev reviewed (approved):
YAY for tests!! 🥳

msalo3 commented on lib/qbwc/response/item_site_query_rs.rb:
```diff
@@ -0,0 +1,86 @@
+            created_at: record['TimeCreated'],
+            updated_at: record['TimeModified'],
+            edit_sequence: record['EditSequence'],
+            full_name: full_name(record),
```
I've been adding extra field for things like this so we have extra info in the transforms.
I'd suggest keeping `full_name`, but also adding:
`inventory_item_name: record['ItemInventoryRef']['FullName']` and `assembly_item_name: record['ItemInventoryAssemblyRef']['FullName']`

Also, I like adding `qbe_item_type: record['ItemInventoryRef'] ? 'qbe_inventory' : 'inventory_assembly'`

raysapida replied:
Sounds good! I'll also take a second look in the documentation and see if there are other fields we could pull

msalo3 commented on lib/qbwc/response/item_site_query_rs.rb:
```diff
@@ -0,0 +1,86 @@
+      def inventories_to_flowlink
+        records.map do |record|
+          {
+            list_id: record['ListID'],
```
We also NEED an `id` field here for FlowLink. For normal products, it would be the name (`id: full_name(record)`), but I'm not exactly sure **what** this object is in QBE?

Does the `record['ListID']` refer to this same item every single time??
If not, I think maybe `id: full_name(record) + record['InventorySiteRef']['FullName']` would be the right thing. The item name + the inventory site is unique.

raysapida replied:
Would full_name be the sku in Systum in that case?

raysapida replied:
The list id refers to the same item I believe but I think it would be better to use your idea of combining the full name and inventory site. It would make the id more readable

msalo3 replied:
FullName _might_ be the SKU in Systum. It also might be something more. FullName sometimes could be Number1ParentName:AnotherParentName:SKU in QBE where the last part matches the SKU in Systum.
Thank god we haven't dealt with anything that crazy yet...

msalo3 commented on lib/qbwc/response/item_site_query_rs.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+      end
+
+      def inventory_site(record)
```
We should test this - I'm not sure we WANT item inventory that doesn't have a Site in QBE.

Maybe we can figure out what exactly inventory without a site is/means? If it's something like "general inventory" that hasn't been added to a site yet, I'd guess we could collate it all into 1 "general" inventory object and return it with all the "site specific" objects?

raysapida replied:
Ahh, this nil check was added when I was getting all the Item sites. Now that we're filtering by site name, this should never be nil

raysapida committed:
Update query response to differentiate between inventory and assembly

raysapida committed:
Add an id field to inventories

msalo3 commented on lib/qbwc/response/item_site_query_rs.rb:
```diff
@@ -0,0 +1,90 @@
+        puts 'ItemSitesQueryRs#process'
+
+        receive_configs = config[:receive] || []
+        inventory_params = receive_configs.find { |c| c['inventories'] }
```
You may have already looked, but where does the integration generate the config files needed for these Get requests? 

I was having issues where getting ServiceProducts wasn't finding the right config here because I was looking for `serviceproducts` but I had run the get_products endpoint.

I'm wondering how exactly this get request might generate the config and will we need to look for `c['inventorywithsites']` or something like that.
I think it will be clearer when testing, so I'm approving, but something to keep an eye on I think!

raysapida replied:
Thanks for the tip! I had to update the PR to use `c['inventorywithsites'` so that it would find the correct config object. For your question, it generates the config in the quickbooks_desktop_endpoint.rb file when it invoked `s3_settings.setup`. That method will look for the JSON file and create it if its not in the folder it expects

raysapida committed:
Fix errors found when testing in staging environment

- A require statement is needed for the new request in the
quickbooks_desktop_integration.rb file
- Params need to be passed instead of config for `#site_name`
- Change the config filter in the response from `inventories` to
`inventorywithsites`

raysapida committed:
Change payload key from `inventories` to `inventorywithsites`

- Thought the payload_key would read the payload_type but its splitting
the endpoint path and using that to find the json file to process

raysapida committed:
Explicity set the payload key when using get_inventorywithsites

This section usually depends on the endpoint being named correctly and
matching the their respective response class using the same key. This
metaprogramming is an issue when we're naming the endpoints differently
because we want to keep the endpoint
